test/integration/remote-config.spec.ts

```diff
@@ -0,0 +1,259 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as admin from '../../lib/index';
+import * as chai from 'chai';
+import * as chaiAsPromised from 'chai-as-promised';
+import { deepCopy } from '../../src/utils/deep-copy';
+
+chai.should();
+chai.use(chaiAsPromised);
+
+const expect = chai.expect;
+
+const VALID_PARAMETERS: { [key: string]: admin.remoteConfig.RemoteConfigParameter } = {
+  // eslint-disable-next-line @typescript-eslint/camelcase
+  holiday_promo_enabled: {
+    defaultValue: { useInAppDefault: true },
+    description: 'promo indicator'
+  },
+  // eslint-disable-next-line @typescript-eslint/camelcase
+  welcome_message: {
+    defaultValue: { value: 'welcome text' },
+    conditionalValues: {
+      ios: { value: 'welcome ios text' },
+      andriod: { value: 'welcome andriod text' },
+    },
+  }
+};
+
+const VALID_CONDITIONS: admin.remoteConfig.RemoteConfigCondition[] = [{
+  name: 'ios',
+  expression: 'device.os == \'ios\'',
+  tagColor: 'BLUE'
+},
+{
+  name: 'andriod',
+  expression: 'device.os == \'android\'',
+  tagColor: 'GREEN'
+}];
+
+const INVALID_PARAMETERS: any[] = [null, '', 'abc', 1, true, []];
+const INVALID_CONDITIONS: any[] = [null, '', 'abc', 1, true, {}];
+const INVALID_TEMPLATES: any[] = [{ parameters: {}, conditions: [], etag: '' }, Object()];
+
+let currentTemplate: admin.remoteConfig.RemoteConfigTemplate;
+
+describe('admin.remoteConfig', () => {
+  before(async () => {
+    // obtain the most recent template (etag) to perform operations
+    currentTemplate = await admin.remoteConfig().getTemplate();
+  });
+
+  it('verify that the etag is read-only', () => {
+    expect(() => {
+      (currentTemplate as any).etag = "new-etag";
+    }).to.throw('Cannot set property etag of #<RemoteConfigTemplateImpl> which has only a getter');
+  });
+
+  describe('validateTemplate', () => {
+    it('should succeed with a vaild template', () => {
+      // set parameters and conditions
+      currentTemplate.conditions = VALID_CONDITIONS;
+      currentTemplate.parameters = VALID_PARAMETERS;
+      return admin.remoteConfig().validateTemplate(currentTemplate)
+        .then((template) => {
+          expect(template.etag).matches(/^etag-[0-9]*-[0-9]*$/);
+          expect(template.conditions.length).to.equal(2);
+          expect(template.conditions).to.deep.equal(VALID_CONDITIONS);
+          expect(template.parameters).to.deep.equal(VALID_PARAMETERS);
+        });
+    });
+
+    INVALID_PARAMETERS.forEach((invalidParameter) => {
```

Comment:
Already covered in unit tests?

Reply:
These tests check for parameters in the `template` that is passed to the `validTemplate()` method.
Current unit tests in `remote-config.spec.ts` validates parameters in the `template` responses from the backend API. That validation happens in the `RemoteConfigTemplateImpl` constructor.

Would you like me to move these to unit tests?

Reply:
I think this is easier to cover in unit tests. Let's move these there.